client/src/test/java/io/pravega/client/segment/impl/SegmentInputStreamTest.java

```diff
@@ -401,4 +417,43 @@ public void testReadWithNegativeEndOffset() {
                 -2, SegmentInputStreamImpl.DEFAULT_BUFFER_SIZE);
     }
     
+    
+    @Test
+    public void testOrderer() throws EndOfSegmentException, SegmentTruncatedException {
+        byte[] data = new byte[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+        ByteBuffer wireData = createEventFromData(data);
+        
+        TestAsyncSegmentInputStream fakeNetwork1 = new TestAsyncSegmentInputStream(segment, 1);
+        TestAsyncSegmentInputStream fakeNetwork2 = new TestAsyncSegmentInputStream(segment, 2);
+        TestAsyncSegmentInputStream fakeNetwork3 = new TestAsyncSegmentInputStream(segment, 3);
+        TestAsyncSegmentInputStream fakeNetwork4 = new TestAsyncSegmentInputStream(segment, 4);
+        
+        @Cleanup
+        SegmentInputStreamImpl stream1 = new SegmentInputStreamImpl(fakeNetwork1, 0);
+        @Cleanup
+        SegmentInputStreamImpl stream2 = new SegmentInputStreamImpl(fakeNetwork2, 0);
+        @Cleanup
+        SegmentInputStreamImpl stream3 = new SegmentInputStreamImpl(fakeNetwork3, 0);
+        @Cleanup
+        SegmentInputStreamImpl stream4 = new SegmentInputStreamImpl(fakeNetwork4, 0);
+
+        fakeNetwork2.complete(0, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, false, wireData.slice()));
+        fakeNetwork3.complete(0, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, true, wireData.slice()));
+        fakeNetwork4.complete(0, new WireCommands.SegmentRead(segment.getScopedName(), 0, false, true, ByteBuffer.allocate(0)));
+        
+        Orderer o = new Orderer();
+        List<SegmentInputStreamImpl> segments = ImmutableList.of(stream1, stream2, stream3, stream4);
+        assertEquals(stream2, o.nextSegment(segments));
+        assertEquals(stream3, o.nextSegment(segments));
+        assertEquals(stream4, o.nextSegment(segments));
+        assertNotNull(stream2.read());
+        assertEquals(stream3, o.nextSegment(segments));
+        assertEquals(stream4, o.nextSegment(segments));
+        assertNotNull(stream3.read());
+        assertEquals(stream3, o.nextSegment(segments));
+        assertEquals(stream4, o.nextSegment(segments));
```

Comment:
The fact that we are returning segment 4 here again makes me think that the semantics of the orderer is the following. For as long as a sealed segment remains in the list of the orderer, the segment is eligible to be returned. It is the responsibility of the caller to take action and process the sealed segment. Is my understanding correct? If so, can we document it in a javadoc somewhere, possibly in the `Orderer` class?

Reply:
Yes. The orderer will return segments that either have data or are sealed. It will not return segments that are not either.